Tree and TreeItem folding helpers and improvements

FeatherAntennae opened the pull request:
A few small changes to the Tree and TreeItem classes.

The goal is to reduce clutter in many classes that each have their own (different) implementation of the same logic, while giving new and old behaviors of the UI trees better consistency across the editor.

**For the discussion and details, see** #33984.

This was separated from the #33984 PR, as the refactor is pretty big and is going to take a while to complete, but I think I pretty much nailed down the global changes needed to make everything compatible and consistent. I think some of these changes would help some other propositions and PR too. #33984 will be kept just to keep track of the refactor, but separating this will allow each part of the refactor to be it's own PR.

I completed the Tree and TreeItem documentation, and opened this PR for the changes to the Tree and TreeItem, including doc for the new methods. It already makes the click, alt-click and folding consistent to all trees by default.

FeatherAntennae commented:
First step in fixing #35529

HaSa1002 commented on doc/classes/TreeItem.xml:
```diff
@@ -411,6 +418,20 @@
+			<argument index="2" name="ignore_self" type="bool">
+			</argument>
+			<description>
+				If [code]enable[/code] is [code]true[/code] collapses, else uncollapses, this TreeItem and all its descendants recursively, ignoring any active TreeItem if [code]ignore_active[/code] is [code]true[/code].
```
This TreeItem and all its descendants collapse recursively when [code]enable[/code] is [code]true[/code], ignoring any active TreeItem if [code]ignore_active[/code] is [code]true[/code].

FeatherAntennae replied:
I think this could be a bit confusing as someone might interpret the "enable" parameter to be a toggle for "recursive or not" instead of a toggle for "collapsed or not".

How about:
>This TreeItem and all its descendants recursively collapse when [code]enable[/code] is [code]true[/code], otherwise expanding, ignoring any active TreeItem if [code]ignore_active[/code] is [code]true[/code].

HaSa1002 commented on doc/classes/TreeItem.xml:
```diff
@@ -411,6 +418,20 @@
+			</argument>
+			<description>
+				If [code]enable[/code] is [code]true[/code] collapses, else uncollapses, this TreeItem and all its descendants recursively, ignoring any active TreeItem if [code]ignore_active[/code] is [code]true[/code].
+				If [code]ignore_self[/code] is [code]true[/code], only affects descendants.
```
Only affects descendants when [code]ignore_self[/code] is [code]true[/code].

FeatherAntennae replied:
I think this could lead to confusion too. It could be interpreted as "descendants are not affected if ignore self is false".

How about:
> Only descendants are affected when [code]ignore_self[/code] is [code]true[/code].

HaSa1002 commented on scene/gui/tree.cpp:
```diff
@@ -347,6 +350,48 @@ void TreeItem::set_collapsed(bool p_collapsed) {
+}
+
+void TreeItem::toggle_collapsed_all_descendants() {
+	if (is_collapsed()) {
+		// Expand recursively.
+		set_collapsed_recursive(false, false, false);
+	} else {
```
I would write
```cpp
if (is_collapsed()) {
    set_collapsed_recursive(false, false, false);
    return;
}
//else block without else here
```

HaSa1002 commented on scene/gui/tree.cpp:
```diff
@@ -347,6 +350,48 @@ void TreeItem::set_collapsed(bool p_collapsed) {
+
+void TreeItem::toggle_collapsed_all_descendants() {
+	if (is_collapsed()) {
+		// Expand recursively.
```
Remove the comment please. The function below tells us that

FeatherAntennae replied:
Will do!

HaSa1002 commented on scene/gui/tree.cpp:
```diff
@@ -1765,7 +1842,12 @@ int Tree::propagate_mouse_event(const Point2i &p_pos, int x_ofs, int y_ofs, bool
 		if (!p_item->disable_folding && !hide_folding && !p_item->cells[col].editable && !p_item->cells[col].selectable && p_item->get_children()) {
-			p_item->set_collapsed(!p_item->is_collapsed());
+			// ALT or equivalent is held, change collapse behavior.
+			if (Input::get_singleton()->is_key_pressed(KEY_ALT)) {
```
I am not sure hardcoding the ALT key here is a good idea, but I don't know if it is done elsewhere this way. I would rather use a method that allows customization.

HaSa1002 replied:
Thinking twice about it, maybe leave it as it is, when the ALT is forced currently (don't know)

FeatherAntennae replied:
I wasn't sure about it either at first, but at the same time, I'm literally coding the "alternate click" function of the little collapse/expand arrow and we already do hardcode the normal click itself.

The few trees in the editor that had a similar function were also hardcoding a key, but some were using ctrl, shift, alt or combination of them to do it.

ALT + Click is definitely the most widespread and standard combination for this kind of behaviour though.

HaSa1002 commented on scene/gui/tree.cpp:
```diff
@@ -3912,6 +3997,12 @@ bool Tree::get_allow_reselect() const {
+void Tree::set_collapsed_all(bool p_collapsed, bool p_ignore_active) {
+
+	get_root()->set_collapsed_recursive(p_collapsed, p_ignore_active, false);
+	ensure_cursor_is_visible();
```
Why needs the cursor to be always visible? I could call that method programmatically where I wouldn't want the cursor to be visible

FeatherAntennae replied:
The name is confusing, but this method was already there and is not really about the cursor itself, but rather what's underneath the cursor. It's actually about the selected TreeItem.

If I don't call "ensure_cursor_is_visible()" when unfolding, the TreeItem that was selected by the user might end up out of view. If the tree is big and requires scrolling both horizontally and vertically, it could end up scrolling the view completely to the right and down or anywhere else depending on the last operations the user did before collapsing everything. 

You normally want the item the user interacted with last to stay visible. 

I usually would agree though, I don't like enforcing things like that, but the goal of this method is to make the "Collapse All" logic the same everywhere across the editor so the UX stays consistent. 

The main use case for this method would be: 
RightClick a TreeItem, which selects it and open a context menu -> select in the menu "Collapse All" or "Extend All". 

Here's a quick example without ensure_cursor_is_visible():
![Ensure Cursor Visible False](https://user-images.githubusercontent.com/52422077/82078846-d37f4d00-96af-11ea-9aed-f39eacb932ca.gif)

And here's with ensure_cursor_is_visible():
![Ensure Cursor Visible True](https://user-images.githubusercontent.com/52422077/82079746-5a80f500-96b1-11ea-9c4c-0c482dcc844b.gif)

I'm open to changing it if needed, but if we do, maybe we should instead add a third boolean property like "skip_ensure_cursor_is_visible" so devs automatically know it's **supposed** to be the default behaviour?

What do you think?

HaSa1002 replied:
So if I understood correctly, the method is all about ensuring the selected item is still in view. (Well as stated in the class reference...). I wouldn't add a third parameter to disable the behaviour, since I can't imagine a case, where it is desired to end up scrolled anywhere.

HaSa1002 replied:
FYI https://github.com/godotengine/godot/issues/16863#issuecomment-629639167

HaSa1002 reviewed (commented):
Just a few I thinks I saw

HaSa1002 commented:
You should squash the last commit. But maybe even meld all of them into one, since they are all used to implement "Add Tree and TreeItem folding helpers" and add the details in the message body

akien-mga commented:
Sorry for the long delay reviewing this. This would likely need to rebased / retested on latest `master` to confirm that it still does what it should.

Would be good to get some input from @godotengine/gui-nodes folks familiar with the Tree API.

> The goal is to reduce clutter in many classes that each have their own (different) implementation of the same logic, while giving new and old behaviors of the UI trees better consistency across the editor.

In this PR I don't see a reduction of said clutter, I assume this was meant to be PR'ed as a follow-up once agreed upon? It would be interesting to see how much these new methods help reduce clutter to be able to assess whether adding them is a good idea or not.

akien-mga commented:
As a matter of fact, I just came to this PR while triaging the least recently updated ones, but we just got a new PR today that seems to address some of the same ideas: #62666. Would be good to have a cross-review @AThousandShips.

FeatherAntennae commented:
> In this PR I don't see a reduction of said clutter, I assume this was meant to be PR'ed as a follow-up once agreed upon? It would be interesting to see how much these new methods help reduce clutter to be able to assess whether adding them is a good idea or not.

The main PR #33984 has more details. This one was split following discussion, to make it easier to review the changes to the `Tree` and `TreeItem` classes. I was a beginner in C++ back then and needed to know if I was breaking anything. In the other PR's description, there are a few of the areas that would have been changed once this one was approved. I would need to go through the codebase again to see where it would be needed now.

> As a matter of fact, I just came to this PR while triaging the least recently updated ones, but we just got a new PR today that seems to address some of the same ideas: https://github.com/godotengine/godot/pull/62666. Would be good to have a cross-review @AThousandShips.

From what I can see, the idea is very similar. The main two differences are :
- I used `ALT + <click on folding arrow>` instead of `SHIFT + <click on folding arrow>`.  
  I went with ALT, as at the time I found more software using ALT instead of SHIFT for this **ALT**ernative functionality. ;)
- I also added the concept of "currently active" branch/hierarchy.  
  It allowed collapsing all nodes recursively, EXCEPT for the one that is currently "active" (and its parents).
  (The difference between active and selected was meant to be that a right click on a different node would select the new node for operations like copy, cut, delete, etc., but would keep the previously selected node as "active" for other functionalities, like for showing properties in a different panel, showing bounding boxes in the scene view, etc.. The other PR #33984 has a gif for a more visual example.

In any cases, if there's interest in this, I will have some free time this week that I could put on updating this and to look at a few of the other elements mentioned before in both PRs. I've been doing Engine code and C++ professionally for a year and a half now, so it'll be much faster. :)

AThousandShips commented:
I went for shift as that was how Scene Tree does it, my original idea was to use right click before I realized it was already a feature

AThousandShips commented:
My code should be ready to go and is in line with what Scene Tree does but since I missed this PR I'll let you see and leave it up to reviewers to decide which way to go

FeatherAntennae commented:
> I went for shift as that was how Scene Tree does it, my original idea was to use right click before I realized it was already a feature

Shift and alt were both hard-coded in various places, which is part of why I was going for this refactor.

Most of the discussion for this was spread through old issues, PRs and a bit on the old IRC, so I don't know if the editor settled more toward one or the other since.

I'll go through the source code later today to see.

FeatherAntennae committed:
Added helper methods to Tree and TreeItem

Tree class:
- Added "set_collapsed_all(...)" method to allow collapsing or expanding
  the entire tree recursively. (Helper method to ensure it is done the
  same way by everyone, making use of the new recursive methods.)

TreeItem class:
- Added "has_expanded_child()" method to detect if at least one child of
  the TreeItem isn't collapsed(and has at least one child of it's own).

- Added "is_in_active_branch()" method to detect if the TreeItem or at
  least one of its descendent is selected.

- Added "set_collapsed_recursive(...)" method to collapse a TreeItem and
  its children recursively with possibility to ignore self and/or skip
  active items.

Tree behaviour:
- ALT + Click on a TreeItem's folding icon will call the TreeItem's
  method "toggle_collapsed_recursive()".

Refactor:
- Removed Scene Tree Dock custom recursive folding logic. It is now
  handled by the Tree directly.

- Replaced Visual Shader Editor collapse / expand all logic with the one
  now included in TreeItem.

- Replaced Script Editor Error Tree collapse / expand all logic with the
  one now included in TreeItem.

Documentation:
- Updated documentation with all new functions.

- Added documentation for "uncollapse_tree()" function.

FeatherAntennae commented:
> Sorry for the long delay reviewing this. This would likely need to rebased / retested on latest `master` to confirm that it still does what it should.

Alright, I redid all of this PR from scratch, much easier than trying to rebase after two years. Also did a quick refactor of a few places in the editor that were using these functions. All trees now supports them by default too.

ALT + Click has the following behaviour:
- If the TreeItem is already collapsed, expands it and all of its children recursively.
- Else, if the TreeItem is expanded and has at least one expanded child, collapses all children recursively without collapsing the current TreeItem.
- Else, the TreeItem is expanded and all its children are collapsed, so  this collapses the TreeItem.

(Essentially, it loops between the 3 states)

![godot windows tools 64_e13LYVYW3F](https://user-images.githubusercontent.com/52422077/177233122-36ad4edd-c17a-40e7-a0f2-458c899b2781.gif)

It also works everywhere else with a tree:
![godot windows tools 64_qsFO0c2A4D](https://user-images.githubusercontent.com/52422077/177233438-52f6a7cc-c4b4-40b1-9248-9d827bedef8f.gif)
![godot windows tools 64_EQKsFl8MTp](https://user-images.githubusercontent.com/52422077/177233443-af90fd72-bfe3-4cd9-9abd-8e1d270f9b36.gif)

KoBeWi commented:
Needs rebase after #62666, if some changes are still relevant.

FeatherAntennae commented:
> Needs rebase after #62666, if some changes are still relevant.

I do not mind rebasing again to contribute the differences between this and #62666, but the main part of what this was doing is done in that other PR, and it's already the second time this got done and forgotten while waiting for review.

#62666 was a new, completely separate and very similar implementation that poped up, bringing attention to this here again. I rewrote everything since it was too old, except for the filesystem dock as it was more work and we were waiting for feedback on the approach, but the other implementation got merged before this one got the review it was waiting for.

Most of what this does seems to have been implemented by #62666, except for two things that were decided based on discussion in the various issues, proposal, and other merge requests related to this one. I assumed the decision to merge the other PR meant that these differences were not wanted.

What was different about my approach:
- Uses "alt" instead of "shift" to avoid accidentally triggering the recursive folding when someone is trying to select a range of item using "shift+click".
- Has 3 states instead of two  https://github.com/godotengine/godot/pull/36926#issuecomment-1174511800 

If these are wanted I can comme back to this, but I'd prefer this to be closed if it's going to be forgotten again because the differences are too small or because we're in the middle of the 4.0 beta.

KoBeWi commented:
If that's the case then I think this can be closed. Moving the folding to Tree unified and removed duplicate code, so it was a straightforward change. The 2 extras you mentioned might warrant a proposal (and a new PR), because they change behavior.

Thanks for your contribution nonetheless and sorry about the outcome .-.
